broker/src/main/java/io/camunda/zeebe/broker/transport/AsyncApiRequestHandler.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.broker.transport;
+
+import io.camunda.zeebe.broker.Loggers;
+import io.camunda.zeebe.broker.transport.AsyncApiRequestHandler.RequestReader;
+import io.camunda.zeebe.broker.transport.AsyncApiRequestHandler.ResponseWriter;
+import io.camunda.zeebe.scheduler.Actor;
+import io.camunda.zeebe.scheduler.future.ActorFuture;
+import io.camunda.zeebe.transport.RequestHandler;
+import io.camunda.zeebe.transport.ServerOutput;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferReader;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+import org.agrona.DirectBuffer;
+import org.agrona.sbe.MessageDecoderFlyweight;
+import org.slf4j.Logger;
+
+/**
+ * A {@link RequestHandler} that automatically decodes requests and encodes successful and error
+ * responses. Handling requests is asynchronous, use {@link ApiRequestHandler} if handling can be
+ * synchronous.
+ *
+ * @param <R> a {@link RequestReader} that reads the request
+ * @param <W> a {@link ResponseWriter} that writes the response
+ */
+public abstract class AsyncApiRequestHandler<R extends RequestReader<?>, W extends ResponseWriter>
+    extends Actor implements RequestHandler {
+
+  public static final Logger LOG = Loggers.TRANSPORT_LOGGER;
+  private final ErrorResponseWriter errorResponseWriter = new ErrorResponseWriter();
+  private final R requestReader;
+  private final W responseWriter;
+
+  protected AsyncApiRequestHandler(final R requestReader, final W responseWriter) {
+    this.requestReader = requestReader;
+    this.responseWriter = responseWriter;
+  }
+
+  /**
+   * Handles a request. The {@param requestReader} is populated at this point.
+   *
+   * @param partitionId the current partition id
+   * @param requestId the current request id
+   * @param requestReader a {@link R reader} that is already populated with the request data
+   * @param responseWriter a {@link W writer} that can be used to write successful responses
+   * @param errorWriter a {@link ErrorResponseWriter} that can be used to write error responses
+   * @return a {@link ErrorResponseWriter} when the handling failed, or a {@link W ResponseWriter}
+   *     when the handling was successful. Writes to the other writer will be ignored and are not
+   *     sent as a response.
+   */
+  protected abstract ActorFuture<Either<ErrorResponseWriter, W>> handleAsync(
+      int partitionId,
+      long requestId,
+      R requestReader,
+      W responseWriter,
+      ErrorResponseWriter errorWriter);
```

Comment:
I don't understand. `handleAsync` gets the reader and writer passed because otherwise it wouldn't have access to those, they are private in `AsyncApiRequestHandler` and IMO should stay that way because `AsyncApiRequestHandler` is responsible for resetting them.

Reply:
I was thinking that the writer can be created in the concrete subclass. Anyway, it was just a thought. You can ignore it.